src/components/shared/Image.tsx

```diff
@@ -0,0 +1,159 @@
+import * as React from 'react'
+import { useRef, useEffect, forwardRef, useImperativeHandle } from 'react'
+import { useAsync } from 'react-use'
+import Services from '../../extension/service'
+import { Skeleton, SkeletonProps } from '@material-ui/lab'
+export interface ImageProps {
+    children?: never
+    src: string | Blob
+    /**
+     * Where should the request be sent?
+     * `current` means the origin of the current page (which means might be blocked by CSP policy)
+     *
+     * `extension` means the request will be sent in the `chrome-extension://` protocol. You need to have the permission to the target site or the site has configured CORS policy correctly.
+     *
+     * `auto` means this component will try to decide this based on "src" property
+     *
+     * @default auto
+     */
+    origin?: 'current' | 'extension' | 'auto'
+    /**
+     * Which tag should be used to render the image?
+     *
+     * - `img` is ignored when `src` is ArrayBuffer.
+     *
+     * - `img` is ignored when `origin` is `extension`.
+     * @default auto
+     */
+    component?: 'img' | 'canvas'
+    width: number
+    height: number
+    loading?: boolean
+    // usability
+    canvasProps?: React.DetailedHTMLProps<React.CanvasHTMLAttributes<HTMLCanvasElement>, HTMLCanvasElement>
+    imgProps?: React.DetailedHTMLProps<React.ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement>
+    SkeletonProps?: SkeletonProps
+    className?: string
+    style?: React.CSSProperties
+}
+
+type ForwardingRef = {
+    img?: HTMLImageElement | null
+    canvas?: HTMLCanvasElement | null
+}
+
+// TODO support concurrent mode
+/**
+ * This React Component is used to render images in the content script to bypass the CSP restriction.
+ */
+export const Image = forwardRef<ForwardingRef, ImageProps>(function Image(props, outgoingRef) {
+    const { src, loading: propsLoading, canvasProps, imgProps, style, className, SkeletonProps } = props
+    // https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas#Maximum_canvas_size
+    const [height, width] = [Math.min(32767, props.height), Math.min(32767, props.width)]
+    const [origin, component] = resolveMode(props)
+    const canvasRef = useRef<HTMLCanvasElement>(null)
+    const imgRef = useRef<HTMLImageElement>(null)
+    useImperativeHandle(outgoingRef, () => ({ canvas: canvasRef.current, img: imgRef.current }), [
+        canvasRef.current,
+        imgRef.current,
+    ])
+
+    // TODO: handle image loading error
+    const { loading, error, value } = useAsync(
+        async function () {
+            if (propsLoading) return
+            if (component === 'img') return
+            if (typeof src !== 'string') return
+            if (origin === 'current') return fetch(src).then((x) => x.blob())
+            return Services.Helper.fetch(src)
+        },
+        [component, src, origin],
+    )
+    if (error) console.error(error)
```

Comment:
Should not continue if fetch failed.

Reply:
IDK what should be returned in this case.

Reply:
Maybe a default image as a placeholder when the error happens.

Reply:
what should it be?

Reply:
A Maskbook logo with opacity 0.1 which is placed in a light gray background. 🤔

Reply:
We'd better leave it in future